### ISSUE TITLE
Revamped build artifacts and versioning

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/sw-helpers#502]
I wanted to outline a proposal around the build output is for our two different types of projects, get consensus, and then implement the changes.

There are a few other issues out there (#470, #487, #444, #361 to some extent) related to the build output, and they would follow once these details are nailed down.

## CommonJS & node CLI Projects
This includes `sw-cli`, `sw-build`, and `sw-build-webpack-plugin`.

### Proposed `build/` output
No substantial changes—we currently copy over the (unbundled) source to the the `build/` directory as-is, without a distinction between `dev` and `prod`.

### Proposed `package.json` metadata
- `"version":` *version string from lerna*
- `"files": ["build"]`
We currently include `"src"` as well, but that's not needed and can be removed.
- `"main": "build/index.js"`

## Service Worker "Runtime" Projects
This includes code that's executed at runtime within the context of a service worker, most notably `sw-lib`, but also the other smaller projects like `sw-runtime-caching` or `sw-background-sync-queue`.

### Proposed `build/` output
```sh
build/
  importScripts/
    <pkg.name>.dev.v<pkg.version>.js
    <pkg.name>.prod.v<pkg.version>.js
  modules/
    <pkg.name>.dev.v<pkg.version>.mjs
    <pkg.name>.prod.v<pkg.version>.mjs
```

The `build/importScripts/` naming emphasizes that the files in that directory are meant to be consumed by the service worker via `importScripts()`.

The `build/modules/` naming, and the `.mjs` extension, emphasizes that the files there are meant to be consumed as ES2015 modules. I don't know if it's currently important to include the versioning info inside of those filenames, but including it is forward-looking for the time in the future in which a service worker could natively import the modules instead of using `importScripts()`.

Getting the correct version number included in the filenames might pose a challenge, as it could involve hooking into the `lerna` build and doing a renaming after `Rollup` bundling is complete but before `lerna` pushes things out to `npm`. I need to explore that more.

Assume that sourcemaps are published alongside each JavaScript bundle.

### Proposed `package.json` metadata
- `"version":` *version string from lerna*
- `"files": ["build"]`
We currently include `"src"` as well, but that's *probably* not needed and can be removed.
- `"main": "build/importScripts/<pkg.name>.prod.v<pkg.version>.js"`
Point to the `prod` `iife` bundle here, primarily because some JS CDNs use that info to determine what gets served by default.
- `"module": "build/modules/<pkg.name>.prod.v<pkg.version>.mjs"`
It looks like the current ES2015 module bundlers have [standardized](https://github.com/rollup/rollup#publishing-es6-modules) on using `module` for this purpose. We can remove `jsnext:main` which we were previously using.

**Comment by jeffposnick** [comment]
(I've edited my initial proposal to add in a `main:dev` field in `package.json`.)

**Comment by gauntface** [comment]
My only concern with not including the version in the dev build is that it adds friction to switching to the production build.

In my head I was thinking developers could change environment by simply replacing 'dev' and 'prod' in the file name, now it's 'dev' and  'prod.<Other>' which I think would actually cause more pain than it solves.

For testing we could create a helper endpoint that would take /__test/find-build/packages/sw-lib/build/sw-lib.dev.js and that return file for /packages/sw-lib/build/sw-lib.dev.v0.0.0.js

**Comment by jeffposnick** [comment]
Agreed that we can handle the automated testing use case via redirects in our test server. Getting the demos happy with URLs that change with every release is a bit trickier.

In any case, it does make more sense to optimize for developer best practices rather than making first-party demos easier to publish, so I can go back to including version info in `dev`.

Maybe the solution is to use template variables in the demos to automatically insert the proper URLs inside of `importScripts()` instead of hardcoding them.

**Comment by addyosmani** [comment]
> Maybe the solution is to use template variables in the demos to automatically insert the proper URLs inside of importScripts() instead of hardcoding them.

That sounds like it should address the concerns around versioning for dev.

**Comment by jeffposnick** [comment]
This was fixed by a combination of #505 and #510
